File: app/src/main/java/com/chatai/adr/repository/ModelManager.java
package com.chatai.adr.repository;

import android.content.Context;
import android.content.SharedPreferences;
import com.chatai.adr.model.LocalModelItem;
import java.io.File;
import java.util.List;

public class ModelManager {

    private static final String PREF_NAME = "model_manager_prefs";
    private static final String KEY_ACTIVE_MODEL_ID = "active_model_id";

    private final Context context;
    private final SharedPreferences prefs;
    private final File modelsDir;
    private final List<LocalModelItem> catalog;

    public interface DownloadProgressListener {
        void onProgress(int progressPercent);
        void onComplete(File downloadedFile);
        void onError(String error);
    }

    public ModelManager(Context context) {
        this.context = context.getApplicationContext();
        this.prefs = this.context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE);
        this.modelsDir = this.context.getDir("llm_models", Context.MODE_PRIVATE);
        this.catalog = ModelCatalog.getRecommendedModels();
        refreshDownloadStatuses();
    }

    public List<LocalModelItem> getCatalog() {
        refreshDownloadStatuses();
        return catalog;
    }

    public void refreshDownloadStatuses() {
        for (LocalModelItem item : catalog) {
            item.updateDownloadStatus(modelsDir);
        }
    }

    public File getModelFile(LocalModelItem item) {
        return new File(modelsDir, item.getFileName());
    }

    public File getModelDirectory() {
        return modelsDir;
    }

    public String getActiveModelId() {
        return prefs.getString(KEY_ACTIVE_MODEL_ID, "qwen2.5-0.5b-instruct");
    }

    public void setActiveModelId(String modelId) {
        prefs.edit().putString(KEY_ACTIVE_MODEL_ID, modelId).apply();
    }

    public LocalModelItem getActiveModel() {
        String activeId = getActiveModelId();
        for (LocalModelItem item : catalog) {
            if (item.getId().equals(activeId)) {
                return item;
            }
        }
        return catalog.isEmpty() ? null : catalog.get(0);
    }

    public boolean deleteModel(LocalModelItem item) {
        File file = getModelFile(item);
        if (file.exists()) {
            boolean deleted = file.delete();
            item.setDownloaded(!deleted);
            return deleted;
        }
        return false;
    }

    /**
     * Creates a lightweight placeholder weight file (for local demo / test purposes)
     * if the real binary is not yet downloaded, enabling instant offline testing.
     */
    public boolean createDemoWeightsIfAbsent(LocalModelItem item) {
        File file = getModelFile(item);
        if (!file.exists()) {
            try {
                if (!modelsDir.exists()) modelsDir.mkdirs();
                boolean created = file.createNewFile();
                item.setDownloaded(true);
                return created;
            } catch (Exception e) {
                return false;
            }
        }
        item.setDownloaded(true);
        return true;
    }
}

File: app/src/main/java/com/chatai/adr/repository/ModelManager.kt
package com.chatai.adr.repository

import android.content.Context
import android.content.SharedPreferences
import android.os.Handler
import android.os.Looper
import com.chatai.adr.model.LocalModelItem
import com.chatai.adr.model.ModelStatus
import java.io.File
import java.io.FileOutputStream

class ModelManager(private val context: Context) {

    private val prefs: SharedPreferences = context.getSharedPreferences(PREFS_MODELS, Context.MODE_PRIVATE)
    private val modelsDir: File = File(context.filesDir, "models").apply { if (!exists()) mkdirs() }
    private val mainHandler = Handler(Looper.getMainLooper())

    var activeModelId: String
        get() = prefs.getString(KEY_ACTIVE_MODEL_ID, "qwen2.5-coder-0.5b") ?: "qwen2.5-coder-0.5b"
        set(value) = prefs.edit().putString(KEY_ACTIVE_MODEL_ID, value).apply()

    fun getRecommendedModels(): List<LocalModelItem> {
        val models = ModelCatalog.recommendedModels
        for (item in models) {
            val file = File(modelsDir, item.localFileName)
            if (file.exists() && file.length() > 0) {
                item.status = ModelStatus.READY
                item.downloadProgress = 100
            } else {
                item.status = ModelStatus.NOT_DOWNLOADED
                item.downloadProgress = 0
            }
        }
        return models
    }

    fun getActiveModel(): LocalModelItem {
        val list = getRecommendedModels()
        return list.firstOrNull { it.id == activeModelId } ?: list.first()
    }

    fun getActiveModelFilePath(): String? {
        val active = getActiveModel()
        val file = File(modelsDir, active.localFileName)
        return if (file.exists() && file.length() > 0) file.absolutePath else null
    }

    fun isAnyModelDownloaded(): Boolean {
        return getRecommendedModels().any { it.isDownloaded }
    }

    fun deleteModel(modelId: String): Boolean {
        val list = getRecommendedModels()
        val item = list.firstOrNull { it.id == modelId } ?: return false
        val file = File(modelsDir, item.localFileName)
        return if (file.exists()) file.delete() else true
    }

    fun simulateDownload(modelId: String, onProgress: (progress: Int, isComplete: Boolean) -> Unit) {
        val list = getRecommendedModels()
        val item = list.firstOrNull { it.id == modelId } ?: return

        item.status = ModelStatus.DOWNLOADING
        item.downloadProgress = 0

        val runnable = object : Runnable {
            var progress = 0
            override fun run() {
                progress += 10
                if (progress > 100) progress = 100
                item.downloadProgress = progress

                if (progress >= 100) {
                    item.status = ModelStatus.READY
                    try {
                        val file = File(modelsDir, item.localFileName)
                        FileOutputStream(file).use { fos ->
                            fos.write("LOCAL_LLM_WEIGHTS_MOCK_HEADER_${item.id}".toByteArray())
                        }
                    } catch (ignored: Exception) {}
                    onProgress(100, true)
                } else {
                    onProgress(progress, false)
                    mainHandler.postDelayed(this, 120)
                }
            }
        }
        mainHandler.post(runnable)
    }

    companion object {
        private const val PREFS_MODELS = "ChatAiModelsPrefs"
        private const val KEY_ACTIVE_MODEL_ID = "active_model_id"
    }
}
